stake-pool/program/src/instruction.rs

```diff
@@ -421,6 +421,58 @@ pub enum StakePoolInstruction {
         #[allow(dead_code)]
         uri: String,
     },
+
+    /// (Staker only) Redelegate active stake on a validator, eventually moving it to another
+    ///
+    /// Internally, this instruction splits a validator stake account into its
+    /// corresponding transient stake account, redelegates it to an ephemeral stake
+    /// account, then merges that stake into the destination transient stake account.
+    ///
+    /// In order to rebalance the pool without taking custody, the staker needs
+    /// a way of reducing the stake on a stake account. This instruction splits
+    /// some amount of stake, up to the total activated stake, from the canonical
+    /// validator stake account, into its "transient" stake account.
+    ///
+    /// The instruction only succeeds if the source transient stake account and
+    /// ephemeral stake account do not exist.
+    ///
+    /// The amount of lamports to move must be at least twice rent-exemption
+    /// plus the minimum delegation amount. Rent-exemption is required for the
+    /// source transient stake account, and rent-exemption plus minimum delegation
+    /// is required for the destination ephemeral stake account.
+    ///
+    ///  0. `[]` Stake pool
+    ///  1. `[s]` Stake pool staker
+    ///  2. `[]` Stake pool withdraw authority
+    ///  3. `[w]` Validator list
+    ///  4. `[w]` Source canonical stake account to split from
+    ///  5. `[w]` Source transient stake account to receive split and be redelegated
+    ///  6. `[w]` Uninitialized ephemeral stake account to receive redelegation
+    ///  7. `[w]` Destination transient stake account to receive ephemeral stake by merge
+    ///  8. `[w]` Destination stake account to receive transient stake after activation
+    ///  9. `[]` Destination validator vote account
+    /// 10. `[]` Clock sysvar
+    /// 11. `[]` Rent sysvar
+    /// 12. `[]` Stake History sysvar
+    /// 13. `[]` Stake Config sysvar
+    /// 14. `[]` System program
+    /// 15. `[]` Stake program
+    Redelegate {
+        /// Amount of lamports to redelegate
+        #[allow(dead_code)] // but it's not
+        lamports: u64,
+        /// Seed used to create source transient stake account
+        #[allow(dead_code)] // but it's not
+        source_transient_stake_seed: u64,
+        /// Seed used to create destination ephemeral account.
+        #[allow(dead_code)] // but it's not
+        ephemeral_stake_seed: u64,
+        /// Seed used to create destination transient stake account. If there is
+        /// already transient stake, this must match the current seed, otherwise
+        /// it can be anything
```

Comment:
Should we recommend the client generate a random u64 value?
`otherwise it can be anything` will cause every good lazy programmer to hard code 0

Reply:
They can be reused, similar to transient stake accounts used for increase / decrease. `0` is fine for the most part, but the stake bots just increment every time to be safe.